python/pyspark/sql/datasource.py

```diff
@@ -0,0 +1,168 @@
+
+__all__ = ["DataSource", "DataSourceReader"]
+
+
+class DataSource:
+    def __init__(self, options: Dict[str, "OptionalPrimitiveType"]):
+        """
+        Initializes the data source with user-provided options.
+
+        Parameters
+        ----------
+        options : dict
+            A dictionary representing the options for this data source.
+
+        Notes
+        -----
+        This method should not contain any non-serializable objects.
+        """
+        self.options = options
+
+    @property
+    def name(self) -> str:
+        """
+        Returns a string represents the short name of this data source.
+        """
+        return self.__class__.__name__
+
+    def schema(self) -> Union[StructType, str]:
+        """
+        Returns the schema of the data source.
+
+        It can reference the `options` field to infer the data source's schema when
+        users do not explicitly specify it. This method is invoked once when calling
+        `spark.read.format(...).load()` to get the schema for a data source read operation.
+        If this method is not implemented, and a user does not provide a schema when
+        reading the data source, an exception will be thrown.
+
+        Returns
+        -------
+        schema : StructType or str
+            The schema of this data source or a DDL string represents the schema
+
+        Examples
+        --------
+        Returns a DDL string:
+
+        >>> def schema(self):
+        >>>    return "a INT, b STRING"
+
+        Returns a StructType:
+
+        >>> def schema(self):
+        >>>   return StructType().add("a", "int").add("b", "string")
+        """
+        ...
+
+    def reader(self, schema: StructType) -> "DataSourceReader":
+        """
+        Returns a DataSourceReader instance for reading data.
+
+        This method is required for readable data sources. It will be called once during
+        the physical planning stage in the Spark planner.
+
+        Parameters
+        ----------
+        schema : StructType
+            The schema of the data to be read.
+
+        Returns
+        -------
+        reader : DataSourceReader
+            A reader instance for this data source.
+        """
+        raise NotImplementedError
+
+
+class DataSourceReader(ABC):
+    """
+    A base class for data source readers. Data source readers are responsible for
+    outputting data from a data source.
+    """
+
+    def partitions(self) -> Iterator[Any]:
+        """
+        Returns a list of partitions for this data source.
+
+        This method is called once during the physical planning stage in the Spark planner
+        to generate a list of partitions. Note, partition values must be serializable.
+
+        The planner then creates an RDD for each partition. Each partition value will be
+        passed to `read(partition)` to read the data from this data source.
+
+        If this method is not implemented, or returns an empty list, Spark will create one
```

Comment:
I believe there is a usecase. For example, in Scala, we have `LocalScan` that doesn't require partitions.

Reply:
Yes, it's possible, but you can always make a UDTF for that use case. That's very simple, you just have to make a class like this:

```
class MyDataSource:
  def eval(self, arg, other):
    # yield rows here
```

The user-defined data source can also explicitly define exactly one partition.
Both options are easy and worth preventing the performance harm from not partitioning a large data source. Having seen the latter before, it often makes the query fail completely.

Reply:
The default partition value should be `None` if the method is not implemented (indicating 1 partition).  I am opting to make `def partitions()` **optional** to implement from a usability perspective, making it easier for new users to get started.

Reply:
If you want to do that, please comment very carefully in the `DataSourceReader` class docstring itself that performance is likely to be limited for large data sets if the `partitions` method is not implemented or returns one partition (but it is OK to skip it if the result data set is small). We will want to make sure that every reader of this abstract base class thinks about this in order to prevent unintentionally making large data sources without any parallelism.